sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java

```diff
@@ -121,24 +141,43 @@ public void processElement(ProcessContext c) throws Exception {
     String tempFilePrefix = resolveTempLocation(
         c.getPipelineOptions().getTempLocation(), "BigQueryWriteTemp", stepUuid);
     TableRowWriter writer = writers.get(c.element().getKey());
+    if (writer != null && writer.getByteSize() > Write.MAX_FILE_SIZE) {
+      // File is too big. Close it and open a new file.
+      TableRowWriter.Result result = writer.close();
+      c.output(new Result<>(result.resourceId.toString(), result.byteSize, c.element().getKey()));
+      writer = null;
+    }
     if (writer == null) {
-      writer = new TableRowWriter(tempFilePrefix);
-      writer.open(UUID.randomUUID().toString());
-      writers.put(c.element().getKey(), writer);
-      LOG.debug("Done opening writer {}", writer);
+      // Only create a new writer if we have fewer than maxNumWritersPerBundle already in thi
+      // bundle.
+      if (writers.size() <= maxNumWritersPerBundle) {
+        writer = new TableRowWriter(tempFilePrefix);
+        writer.open(UUID.randomUUID().toString());
+        writers.put(c.element().getKey(), writer);
+        LOG.debug("Done opening writer {}", writer);
+      }
     }
-    try {
-      writer.write(c.element().getValue());
-    } catch (Exception e) {
-      // Discard write result and close the write.
+    if (writer != null) {
       try {
-        writer.close();
-        // The writer does not need to be reset, as this DoFn cannot be reused.
-      } catch (Exception closeException) {
-        // Do not mask the exception that caused the write to fail.
-        e.addSuppressed(closeException);
+        writer.write(c.element().getValue());
+      } catch (Exception e) {
+        // Discard write result and close the write.
+        try {
+          writer.close();
+          // The writer does not need to be reset, as this DoFn cannot be reused.
+        } catch (Exception closeException) {
+          // Do not mask the exception that caused the write to fail.
+          e.addSuppressed(closeException);
+        }
+        throw e;
       }
-      throw e;
+    } else {
+      // This means that we already had too many writers open in this bundle. "spill" this record
+      // into the output. It will be grouped and written to a file in a subsequent stage.
+      c.output(unwrittedRecordsTag,
+          KV.of(ShardedKey.of(c.element().getKey(),
+              ThreadLocalRandom.current().nextInt(SPILLED_RECORD_SHARDING_FACTOR)),
+              c.element().getValue()));
     }
   }
 
```

Comment:
In FinishBundle, we might get an exception from c.output(), or from a writer's close(). In that case, the later writers might not get closed.